### PR TITLE
Adding the ability to set API versions at the REST request layer

### PR DESCRIPTION
**Pull request by bhariharan**
Currently, a consuming app can only set the API version at the REST API layer, which sets it for all requests coming through. API version should be set at the REST request layer - this gives a consuming app the ability to specify which version to use in a more granular manner. This PR is a first stab at allowing for this. This feature already exists on `Android`.

Outstanding Work:

* Replace deprecated method usage in `SmartSync` and test targets with the newer factory methods (will be in the next PR).
* Get rid of deprecated methods (will be in `Mobile SDK 8.0` - @bbirman this one's for you).

### CODE REVIEW

**Comment by sgoldberg-sfdc** [comment]
👍 Thanks for this Bharath

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -3250,6 +3250,7 @@
 				GCC_WARN_UNUSED_VARIABLE = YES;
 				SDKROOT = iphoneos;
 				SUPPORTED_PLATFORMS = "iphonesimulator iphoneos";
+				SWIFT_COMPILATION_MODE = wholemodule;
```
Comment (bbirman) [inline libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: What does this do?

Comment (bhariharan) [reply libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: Good question - that was the result of clicking `Update to Xcode recommended settings` in `Xcode 10.3`.

**Comment by trooper2013** [comment]
@bhariharan Weren't you going to make the api version Nullable?  i.e. the default will be used if nil.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI+Files.h [hunk 2]
```diff
@@ -37,99 +37,209 @@ NS_ASSUME_NONNULL_BEGIN
...
+
+/**
+ * Build a request that will delete the specified file share.
+ *
+ * @param shareId The Id of the file share record (aka ContentDocumentLink).
+ * @param apiVersion API version.
+ * @return A new SFRestRequest that can be used to delete this share.
  */
-- (SFRestRequest *) requestForDeleteFileShare:(NSString *)shareId;
+- (SFRestRequest *)requestForDeleteFileShare:(NSString *)shareId apiVersion:(NSString *)apiVersion;
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI+Files.h]: Should we make the version Nullable in all calls?

**Comment by bhariharan** [comment]
@trooper2013 I'll do that in the next PR. I'm planning on moving the constants to a class and plugging them in.

**Comment by bhariharan** [comment]
@trooper2013 Merging this for now - will add more in subsequent PRs, so that it's easier to review.

**Pull request merged**